### ISSUE TITLE
RProp

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#7603]
RProp<br>
<br>
https://ift.tt/3sHtPYW<br>
<br>
About 0 Minutes<br>
<br>
<div>
<article>
<p><em>RProp</em> is a popular gradient descent algorithm that only uses the signs of gradients to compute updates <a href="#citation-1">[1]</a> <a href="#citation-2">[2]</a> . It stands for <em>Resilient Propagation</em> and works well in many situations because it adapts the step size dynamically for each weight independently. This blog posts gives an introduction to RProp and motivates its design choice of ignoring gradient magnitudes.</p>
<p>Most gradient descent variants use the sign and the magnitude of the gradient. The gradient points in the direction of steepest ascent. Because we typically want to find a minimum, we follow the gradient in the opposite direction. This direction is completely determined by the sign of the gradient.</p>
<h3>Gradient magnitudes<a href="#gradient-magnitudes"></a>
</h3>
<p>To decide on the step size, a scaled version of the gradient&rsquo;s magnitude is generally used by most gradient descent algorithms. This heuristic often works well but there is no guarantee that it is always a good choice. To see that it can work extremely badly, and does not have to contain valuable information, we consider a function \(f\). The plots below show \(f\) as well as two scaled versions.</p>
<table>
<caption align="bottom">Three functions with the same optima but vastly different gradients</caption>
<tbody>
<tr>
<td><img alt="Three functions with the same optima but vastly different gradients" src="https://florian.github.io/assets/posts/rprop/scales.png"></td>
</tr>
</tbody>
</table>
<p>All three of these functions have the exact same optima, so the step updates using gradient descent should all be similar. However, if we determine the step size using the gradient&rsquo;s magnitude, then the step sizes for the three functions differ by orders of magnitude. Even worse, the gradient virtually vanishes for the second function and explodes for the third, as shown below:</p>
<table>
<caption align="bottom">The first derivatives of the three functions</caption>
<tbody>
<tr>
<td><img alt="The first derivatives of the three functions " src="https://florian.github.io/assets/posts/rprop/scales-diff.png"></td>
</tr>
</tbody>
</table>
<p>This shows that the gradient&rsquo;s magnitude does not necessarily contain useful information for determining the step size. Even though optima can still be found by choosing appropriate learning rates, this makes it clear that using the gradient&rsquo;s magnitude at all is sometimes questionable. Using a fixed learning rate will also fail if only some parts of the function are scaled.</p>
<h3>Updating weights<a href="#updating-weights"></a>
</h3>
<p>Modern gradient descent variants try to circumvent this problem by dynamically adapting the step size. RProp does this in a way that only requires the sign of the gradient. By ignoring the gradient&rsquo;s magnitude, RProp has no problems if a function has a few very steep areas.</p>
<p>Concretely, RProp uses a different step size for each dimension. Let \(\eta_i^{(t)}\) be the step size for the \(i\)-th weight in the \(t\)-th iteration of gradient descent. The value for the first and second iteration, \(\eta_i^{(0)}\) and \(\eta_i^{(1)}\), is a hyperparameter that needs to be chosen in advance. This step size is then dynamically adapted for each weight, depending on the gradient.</p>
<p>The weights themselves are updated using</p>
\[w_i^{(t)} = w_i^{(t - 1)} - \eta_i^{(t - 1)} * \operatorname{sgn}\left(\frac{\partial E^{(t - 1)}}{\partial w_i^{(t - 1)}}\right)\]
<p>where the sign of the partial derivative of the error in the last step with respect to the respective weight is computed. We go in the direction of descent using the determined step size.</p>
<h3>Adapting the step size<a href="#adapting-the-step-size"></a>
</h3>
<p>In each iteration of RProp, the gradients are computed and the step sizes are updated for each dimension individually. This is done by comparing the gradient&rsquo;s sign of the current and previous iteration. The idea here is the following:</p>
<ul>
<li>When the signs are the same, we go in the same direction as in the previous iteration. Since this seems to be a good direction, the step size should be increased to go to the optimum more quickly</li>
<li>If the sign changed, the new update is moving in a different direction. This means that we just jumped over an optimum. The step size should be decreased to avoid jumping over the optimum again</li>
</ul>
<p>A visualization of this idea is shown below.</p>
<table>
<caption align="bottom">The gradient direction changes when jumping over optima</caption>
<tbody>
<tr>
<td><img alt="The gradient direction changes when jumping over optima" src="https://florian.github.io/assets/posts/rprop/jumps.png"></td>
</tr>
</tbody>
</table>
<p>To implement this update scheme, the following formula is used:</p>
\[\eta_i^{(t)} = \begin{cases} \min(\eta_i^{(t - 1)} * \alpha, \eta_{\max}) &amp; \text{if } \frac{\partial E^{(t)}}{\partial w_i^{(t)}} * \frac{\partial E^{(t - 1)}}{\partial w_i^{(t - 1)}} &gt; 0 \\ \max(\eta_i^{(t - 1)} * \beta, \eta_{\min}) &amp; \text{if } \frac{\partial E^{(t)}}{\partial w_i^{(t)}} * \frac{\partial E^{(t - 1)}}{\partial w_i^{(t - 1)}} &lt; 0 \\ \eta_i^{(t - 1)} &amp; \text{otherwise} \end{cases} \label{eq:rprop}\]
<p>where \(\alpha &gt; 1 &gt; \beta\) scale the step size, depending on whether the speed should be increased or decreased. The step size is then clipped using \(\eta_{\min}\) and \(\eta_{\max}\) to avoid it becoming too large or too small. If a gradient was zero, a local optimum for this weight was found and the step size is not changed.</p>
<h3>Hyperparameters<a href="#hyperparameters"></a>
</h3>
<p>These seem like many hyperparameters to choose, but in practice there are known values for them that generally work well. It is also not problematic if the clipping values \(\eta_{\min}\) and \(\eta_{\max}\) are respectively smaller and larger than necessary because an inconvenient step size is generally adapted quickly.</p>
<p>Popular values for \(\alpha\) and \(\beta\) are \(1.2\) and \(0.5\). Heuristically, it works well to increase the step size slowly, while allowing for the possibility of quickly decreasing it when jumping around an optimum. For fine-tuning the weights, it is important that \(\beta\) is not the reciprocal of \(\alpha\), to allow for many different step sizes.</p>
<h3>Conclusion<a href="#conclusion"></a>
</h3>
<p>One advantage of RProp that was not discussed so far is having a different step size for each weight. If one weight is already very close to its optimal value while a second weight still needs to be changed a lot, this is not a problem for RProp. Other gradient descent variants can have much more problems with such a situation, especially because the gradient magnitudes can be misleading here.</p>
<p>While RProp works well in a lot of situations, it is not perfect. For instance, RProp generally requires large batch updates. If there&rsquo;s too much randomness in stochastic gradient descent, then the step sizes jump around too much and the updates work badly.</p>
<p>Implementing RProp is quite straightforward. To get a better understanding of RProp, reading the <a href="https://github.com/pytorch/pytorch/blob/master/torch/optim/rprop.py">PyTorch implementation</a> can also be helpful.</p>
<h3>References<a href="#references"></a>
</h3>
<ol>
<li><span>Rojas, R., 2013. Neural networks: a systematic introduction. Springer Science &amp; Business Media. <a href="#ref-1"></a></span></li>
<li><span>Riedmiller, M. and Braun, H., 1993. A direct adaptive method for faster backpropagation learning: The RPROP algorithm. In Neural Networks, 1993., IEEE International Conference on (pp. 586-591). IEEE. <a href="#ref-1"></a></span></li>
</ol>
</article>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via florian.github.io https://ift.tt/3oYyq6p<br>
<br>
January 19, 2021 at 09:24PM
